### DISCUSSION TITLE
How to handle Error with lazy Route: Failed to fetch dynamically imported module

### DISCUSSION THREAD

**Discussion by lasseklovstad** [remix-run/react-router#10333] [Q&A] [answered]
Hi!

I have started using the lazy Route functionality and reduced my build index file by 50% 😄 and thats nice. But I have started seeing some errors that lazy loaded chunk files are not found. This often happens when the app is redeployed as [described here](http://dimaip.github.io/2020/04/25/deploying-apps-with-code-splitting/). 

I would love it if anyone have any recomandations of how to handle this correctly.

I have tried this, but this creates a typeerror, but i guess would potentially work?:
```typescript
const router = createBrowserRouter([
  {
    path: "/",
    lazy: () => import("./routes/Demo/Demo").catch(() => window.location.reload()),
  },
]);
```

**Comment by brophdawg11** [comment, marked as answer]
Errors from `lazy` should bubble to route error boundaries, so you can either force a hard refresh like you're doing above or you could provide a UI to the user with a button to reload:

```jsx
const router = createBrowserRouter([
  {
    path: "/",
    ErrorBoundary() {
      return (
        <>
          <h1>Uh oh!</h1>
          <p>Something went wrong!<p>
          <button onClick={() => window.location.reload()}>Click here to reload the page</button>
        </>
      );
    },   
    lazy: () => import("./routes/Demo/Demo").catch(() => window.location.reload()),
  },
]);
```

The type error is likely just that lazy expects to resolve with a route module, so this might fix it:

```js
lazy: () => import("./routes/Demo/Demo").catch(() => {
  window.location.reload();
  return {};
}),
```

**Reply by User6531** [reply]
Should I create a `catch block` for each route, or can I write some sort of interceptor to catch this error for all routes and then reload the page?

**Reply by brophdawg11** [reply]
A root level error boundary will catch all occurrences if you don't want a `catch` on each one.  Just remember that'll catch _other_ non-import errors as well so you may want to do some filtering on the error before triggering a hard reload.

**Reply by User6531** [reply]
Yeah, exactly. I want to include this hard reload only when the dynamically imported module fails to fetch. So, I'm not quite sure about the exact way to accurately catch this error.

**Reply by User6531** [reply]
I found this 

```ts
if (error.message.includes('Failed to fetch dynamically imported module')) {
    window.location.reload();
}
```

but I don`t like it 🙂

**Reply by lasseklovstad** [reply]
I found out that chrome/Firefox/Safari throws different messages. So I ended up doing this, but yeh it feals hacky:
```typescript
  if (error.message.includes("dynamically imported module")) {
    // TypeError
    // Error message is different in some browsers:
    // Chrome: Failed to fetch dynamically imported module
    // Firefox/Safari: error loading dynamically imported module
```

**Reply by brophdawg11** [reply]
Just use a utility function:

```js
async function importOrReload(dynamicImportFn) {
  try {
    return dynamicImportFn();
  } catch (e) {
    window.location.reload();
  }
}

const router = createBrowserRouter([
  {
    path: "/",
    lazy: importOrReload(() => import("./layout")),
    children: [{
      index:true,
      lazy: importOrReload(() => import("./home")),
    }]
  },
]);
```

**Reply by User6531** [reply]
Thanks a lot for your response. 
I think it might work better if we only refresh the page when we encounter an error specifically related to the `Failed to fetch dynamically imported module` instead of refreshing for every error. 
This way, we can let the error boundary catch and display other types of errors to the user.
Your understanding on this would be really appreciated.

**Reply by brophdawg11** [reply]
This was just an example - you can alter the logic in the utility function however you wish - such as conditionally reloading based on the error.  But the idea here is that the `catch` is scoped to the dynamic import so it should only be catching those specific errors and does not rely on error-prone string matching across browsers.

**Comment by vincerubinetti** [comment]
How can one customize the behavior of error handling this specific "Error loading route module" error in React Router v7 framework mode?
